File: csharp/sum-of-multiples/SumOfMultiples.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class SumOfMultiples
{
    private List<int> factors;

    public SumOfMultiples() : this(new List<int> { 3, 5 }) { }

    public SumOfMultiples(List<int> factors)
    {
        this.factors = factors;
    }

    public int To(int upTo)
    {
        var multiples = new HashSet<int>();
        factors.ForEach(delegate(int factor) 
        { 
            multiples.UnionWith(MultiplesToFor(upTo, factor)); 
        });
        return multiples.Sum<int>(x => x);
    }

    private IEnumerable<int> MultiplesToFor(int upTo, int factor)
    {
        var mcount = (int) Math.Ceiling((double) upTo / factor) - 1;
        return Enumerable.Range(1, mcount).Select(i => i * factor);
    }
}

File: haskell/sum-of-multiples/SumOfMultiples.hs
module SumOfMultiples (sumOfMultiples, sumOfMultiplesDefault) where

import Data.List (any)

isMultiple :: [Integer] -> Integer -> Bool
isMultiple factors x = any (\f -> x `mod` f == 0) factors

sumOfMultiples :: [Integer] -> Integer -> Integer
sumOfMultiples factors upTo = sum [ x | x <- [1.. upTo - 1], isMultiple factors x ] 

sumOfMultiplesDefault :: Integer -> Integer
sumOfMultiplesDefault = sumOfMultiples [3,5]